sql/src/main/java/io/crate/operation/projectors/UpdateProjector.java

```diff
@@ -0,0 +1,193 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.operation.projectors;
+
+import com.google.common.util.concurrent.SettableFuture;
+import io.crate.Constants;
+import io.crate.metadata.ColumnIdent;
+import io.crate.operation.ProjectorUpstream;
+import io.crate.operation.collect.CollectExpression;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.update.TransportUpdateAction;
+import org.elasticsearch.action.update.UpdateRequest;
+import org.elasticsearch.action.update.UpdateResponse;
+import org.elasticsearch.index.mapper.Uid;
+import org.elasticsearch.index.shard.ShardId;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class UpdateProjector implements Projector {
+
+    private Projector downstream;
+    private final AtomicInteger remainingUpstreams = new AtomicInteger(0);
+    private final AtomicReference<Throwable> upstreamFailure = new AtomicReference<>(null);
+    private final AtomicInteger pendingUpdateRequests = new AtomicInteger(0);
+    private final SettableFuture<Long> rowCountFuture = SettableFuture.create();
+
+    private final ShardId shardId;
+    private final TransportUpdateAction transportUpdateAction;
+    private final List<ColumnIdent> columnIdents;
+    private final CollectExpression<?>[] collectExpressions;
+    @Nullable
+    private final Long requiredVersion;
+    private final Object lock = new Object();
+    private final AtomicLong rowCount = new AtomicLong(0);
+    private final ActionListener<UpdateResponse> listener;
+
+    public UpdateProjector(ShardId shardId,
+                           TransportUpdateAction transportUpdateAction,
+                           List<ColumnIdent> columnIdents,
+                           CollectExpression<?>[] collectExpressions,
+                           @Nullable Long requiredVersion) {
+        this.shardId = shardId;
+        this.transportUpdateAction = transportUpdateAction;
+        this.columnIdents = columnIdents;
+        this.collectExpressions = collectExpressions;
+        this.requiredVersion = requiredVersion;
+        listener = new ActionListener<UpdateResponse>() {
+            @Override
+            public void onResponse(UpdateResponse updateResponse) {
+                rowCount.incrementAndGet();
+                if (pendingUpdateRequests.decrementAndGet() == 0) {
```

Comment:
isn't it possible to have one update request get executed (pendingUpdateRequests = 1), get finished (pendingUpdateRequests = 0) --> rowCountFuture.set(...) and then another update request is executed but its rowcount is not considered anymore?